airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueIdentityMapper.kt

```diff
@@ -47,43 +48,60 @@ open class AirbyteValueIdentityMapper(
-                is UnionType -> mapUnion(value, schema, path)
-                is BooleanType -> mapBoolean(value as BooleanValue, path)
-                is NumberType -> mapNumber(value as NumberValue, path)
-                is StringType -> mapString(value as StringValue, path)
-                is IntegerType -> mapInteger(value as IntegerValue, path)
-                is DateType -> mapDate(value as DateValue, path)
-                is TimeTypeWithTimezone -> mapTimeWithTimezone(value as TimeValue, path)
-                is TimeTypeWithoutTimezone -> mapTimeWithoutTimezone(value as TimeValue, path)
-                is TimestampTypeWithTimezone ->
-                    mapTimestampWithTimezone(value as TimestampValue, path)
-                is TimestampTypeWithoutTimezone ->
-                    mapTimestampWithoutTimezone(value as TimestampValue, path)
-                is UnknownType -> {
-                    collectFailure(path)
-                    mapNull(path)
+            if (nullable && value is NullValue) {
+                mapNull(path)
+            } else {
+                when (schema) {
+                    is ObjectType -> mapObject(value as ObjectValue, schema, path)
+                    is ObjectTypeWithoutSchema ->
+                        mapObjectWithoutSchema(value as ObjectValue, schema, path)
+                    is ObjectTypeWithEmptySchema ->
+                        mapObjectWithEmptySchema(value as ObjectValue, schema, path)
+                    is ArrayType -> mapArray(value as ArrayValue, schema, path)
+                    is ArrayTypeWithoutSchema ->
+                        mapArrayWithoutSchema(value as ArrayValue, schema, path)
+                    is UnionType -> mapUnion(value, schema, path)
+                    is BooleanType -> mapBoolean(value as BooleanValue, path)
+                    is NumberType -> mapNumber(value as NumberValue, path)
+                    is StringType -> mapString(value as StringValue, path)
+                    is IntegerType -> mapInteger(value as IntegerValue, path)
+                    is DateType -> mapDate(value as DateValue, path)
+                    is TimeTypeWithTimezone ->
+                        mapTimeWithTimezone(
+                            value as TimeValue,
+                            path,
+                        )
+                    is TimeTypeWithoutTimezone ->
+                        mapTimeWithoutTimezone(
+                            value as TimeValue,
+                            path,
+                        )
+                    is TimestampTypeWithTimezone ->
+                        mapTimestampWithTimezone(value as TimestampValue, path)
+                    is TimestampTypeWithoutTimezone ->
+                        mapTimestampWithoutTimezone(value as TimestampValue, path)
+                    is UnknownType -> mapUnknown(value, path)
                 }
             }
         } catch (e: Exception) {
+            if (!nullable) {
```

Comment:
I think this should be in the `if (nullable && value is NullValue)` branch, instead of in the exception handler:
```kotlin
if (value is NullValue) {
  if (nullable) mapNull(path)
  else throw
}
```

(also, checking: this should never happen for actual record data, b/c the only nonnullable fields are the airbyte_foo fields? I.e. we _want_ to hard crash here, because this indicates that we actually messed up somewhere)

Reply:
Yes, this indicates a programming error

Reply:
I added this btw